File: day08/main.go
package main

import (
	"aocutil"
)

func day8_count_string_memory(str string) (visualLength, memoryLength, encodedLength int) {
	visualLength = len(str)
	encodedLength = 6

	for i := 1; i < len(str)-1; i++ {
		switch str[i] {
		case '\\':
			if i+3 < visualLength && str[i+1] == 'x' {
				if aocutil.IsHexChar(str[i+2]) && aocutil.IsHexChar(str[i+3]) {
					i += 3
					encodedLength += 3
				}
			} else if i+1 < visualLength {
				if str[i+1] == '\\' || str[i+1] == '"' {
					i++
					encodedLength += 2
				}
			}
			encodedLength += 2
			memoryLength++
		default:
			encodedLength++
			memoryLength++
		}
	}

	return
}

func main() {
	visualTotal := 0
	memoryTotal := 0
	encodedTotal := 0

	aocutil.FileReadAllLines("day08/input.txt", func(s string) {
		visual, memory, encoded := day8_count_string_memory(s)
		visualTotal += visual
		memoryTotal += memory
		encodedTotal += encoded
	})

	aocutil.AOCFinish(visualTotal-memoryTotal, encodedTotal-visualTotal)
}

File: day08/main.c
#include <aoc/filesystem.h>
#include <stdio.h>
#include <string.h>
#include <aoc/aoc.h>

typedef struct {
  size_t visual, memory, encoded;
} context;

static void parse(char *line, context *ctx) {
  size_t totalLength = strlen(line);
  aoc_string_trim_right(line, &totalLength);
  ctx->visual += totalLength;
  ctx->encoded += 6; //  "\"\""

  line++; // "
  while (*line != '"') {
    switch (*line) {
    case '\\':
      switch (line[1]) {
      case 'x':
        line += 4;
        ctx->encoded += 4;
        break;
      case '\\':
      case '"':
        line += 2;
        ctx->encoded += 3;
        break;
      }
      break;
    default:
      line++;
      break;
    }
    ctx->memory += 1;
    ctx->encoded += 1;
  }
}

int main(void) {
  context ctx = {0};
  aoc_file_read_lines1("day08/input.txt", (aoc_line_func)parse, &ctx);
  printf("%zu\n", ctx.visual - ctx.memory);
  printf("%zu\n", ctx.encoded - ctx.visual);
}
